File: 백준/Silver/2805. 나무 자르기/나무 자르기.cc
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int main() {
    int Tree_num, need_length;
    cin >> Tree_num >> need_length;

    vector<int> Tree(Tree_num);
    for(int i = 0; i < Tree_num; i++) {
        cin >> Tree[i];
    }
    long long start = 1;
    long long end = *max_element(Tree.begin(), Tree.end());
    int result = 0;
    while(start <= end) {
        long long mid = (start + end) / 2;
        long long get_Tree = 0;
        for(int i = 0; i < Tree_num; i++) {
            if(Tree[i] > mid) // 나무가 잘리는 경우
                get_Tree += Tree[i] - mid;
        }
        if(get_Tree >= need_length) {// 더 많은 나무를 가지게 되는 경우
            start = mid + 1;
        }
        else // 목표치보다 더 작은 나무를 가지게 되는 경우         
            end = mid - 1;
    }

    cout << end << endl;

    return 0;
}

File: 백준/Silver/2805. 나무 자르기/나무 자르기.py
import sys
input = sys.stdin.readline

Tree_num, need_length = map(int, input().split())
Tree = list(map(int, input().split()))

start, end = 1, max(Tree)
while start <= end:
    mid = (start + end) // 2
    get_Tree = 0
    for tr in Tree:
        if tr > mid: # 나무가 잘리는 경우
            get_Tree += tr - mid
    if get_Tree >= need_length: # 더 많은 나무를 가지게 되는 경우
        start = mid + 1
    else: # 목표치보다 더 작은 나무를 가지게 되는 경우
        end = mid - 1
        
print(end)